crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
 pub fn calculate_bounds(
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
-    without_aabb: Query<(Entity, &Handle<Mesh>), (Without<Aabb>, Without<NoFrustumCulling>)>,
+    without_aabb_or_with_changed_mesh: Query<
+        (Entity, &Handle<Mesh>),
+        (
+            Or<(Without<Aabb>, Changed<Handle<Mesh>>)>,
+            Without<NoFrustumCulling>,
+        ),
+    >,
+    mut entity_mesh_map: Local<HashMap<Handle<Mesh>, Vec<Entity>>>,
+    mut mesh_events: EventReader<AssetEvent<Mesh>>,
 ) {
-    for (entity, mesh_handle) in without_aabb.iter() {
+    for (entity, mesh_handle) in without_aabb_or_with_changed_mesh.iter() {
+        // Record entities that have mesh handles.
+        entity_mesh_map
+            .entry(mesh_handle.clone())
+            .or_default()
+            .push(entity);
+
         if let Some(mesh) = meshes.get(mesh_handle) {
             if let Some(aabb) = mesh.compute_aabb() {
                 commands.entity(entity).insert(aabb);
             }
         }
     }
+
+    // Calculate bounds for entities whose meshes have been mutated.
+    let updated_mesh_handles = mesh_events.iter().filter_map(|event| match event {
+        AssetEvent::Modified { handle } => Some(handle),
+        _ => None,
+    });
+    let updated_meshes_and_entities = updated_mesh_handles.filter_map(|mesh_handle| {
+        meshes
+            .get(mesh_handle)
+            .zip(entity_mesh_map.get(mesh_handle))
```

Comment:
This list never gets entities removed from it. So if an entity gets a new handle and then its old handle is mutated we'd set its Aabb to the new geometry of the old mesh. We probably don't want that. Is there an easier way to keep track of these than another `Local<HashMap<Entity, Handle<Mesh>>>`?

Reply:
Ah. This is https://github.com/bevyengine/bevy/issues/4513. I would go with the naive uncached version, and leave a `PERF` comment.

As you've seen, there's a lot of challenges regarding data synchronization when trying to roll your own indexes.

Reply:
Attempted to handle in 5ee488a5e60ff51e30f309b103c6a4f9a12b1606. Updated PR description with an app that maybe shows the bug and the fix.